src/test/scala/com/microsoft/hyperspace/index/BucketUnionTest.scala

```diff
@@ -0,0 +1,126 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.microsoft.hyperspace.index
+
+import org.apache.hadoop.fs.Path
+import org.apache.spark.sql.Row
+import org.apache.spark.sql.catalyst.catalog.BucketSpec
+
+import com.microsoft.hyperspace.index.execution.{BucketUnionExec, BucketUnionRDD, BucketUnionRDDPartition, BucketUnionStrategy}
+import com.microsoft.hyperspace.index.plans.logical.BucketUnion
+
+class BucketUnionTest extends HyperspaceSuite {
+  override val systemPath = new Path("src/test/resources/bucketUnionTest")
+
+  test("BucketUnion require test") {
+    import spark.implicits._
+    val df1 = Seq((1, "name1"), (2, "name2")).toDF("id", "name")
+    val df1_1 = Seq((1, "name1"), (2, "name2")).toDF("id", "name")
+    val df1_2 = Seq(("name1", 1), ("name2", 2)).toDF("name", "id")
+    val df2 = Seq((1, "name1", 20), (2, "name2", 10)).toDF("id", "name", "age")
+
+    // different column schema
+    intercept[IllegalArgumentException] {
+      BucketUnion(
+        Seq(df1.queryExecution.optimizedPlan, df2.queryExecution.optimizedPlan),
+        BucketSpec(1, Seq(), Seq()))
+    }
+
+    // different order of columns
+    intercept[IllegalArgumentException] {
+      BucketUnion(
+        Seq(df1.queryExecution.optimizedPlan, df1_2.queryExecution.optimizedPlan),
+        BucketSpec(1, Seq(), Seq()))
+    }
+
+    BucketUnion(
+      Seq(df1.queryExecution.optimizedPlan, df1_1.queryExecution.optimizedPlan),
+      BucketSpec(1, Seq(), Seq()))
+  }
+
+  test("BucketUnionStrategy test") {
+    import spark.implicits._
+    val df1 = Seq((1, "name1"), (2, "name2")).toDF("id", "name")
+    val df1_1 = Seq((1, "name1"), (2, "name2")).toDF("id", "name")
+    val bucket = BucketUnion(
+      Seq(df1.queryExecution.optimizedPlan, df1_1.queryExecution.optimizedPlan),
+      BucketSpec(1, Seq(), Seq()))
+
+    assert(BucketUnionStrategy(bucket).collect {
+      case BucketUnionExec(_, _) => true
+    }.length == 1)
+
+    assert(BucketUnionStrategy(df1.queryExecution.optimizedPlan).collect {
+      case BucketUnionExec(_, _) => true
+    }.isEmpty)
+  }
+
+  test("BucketUnionExec test") {
+    import spark.implicits._
+    val df1 = Seq((1, "name1"), (2, "name2")).toDF("id", "name")
+    val p1 = df1.repartition(10)
+    val df1_1 = Seq((1, "name1"), (2, "name2")).toDF("id", "name")
+    val p1_1 = df1_1.repartition(9)
+    val p1_2 = df1_1.repartition(10)
+
+    // different number of partition
+    intercept[AssertionError] {
+      val bucket = BucketUnion(
+        Seq(p1.queryExecution.optimizedPlan, p1_1.queryExecution.optimizedPlan),
+        BucketSpec(10, Seq(), Seq()))
+      spark.sessionState.executePlan(bucket).sparkPlan
+    }
+
+    val bucket = BucketUnion(
+      Seq(p1.queryExecution.optimizedPlan, p1_2.queryExecution.optimizedPlan),
```

Comment:
nit: only if you make the change above
```suggestion
      Seq(p1.queryExecution.optimizedPlan, p2_2.queryExecution.optimizedPlan),
```